File: sdks/go/autonoma/payload_topo.go
package autonoma

import (
	"fmt"
	"sort"
	"strings"
)

// reservedKeys are keys in entity objects that the SDK interprets specially.
var reservedKeys = map[string]bool{
	"_alias": true,
	"_ref":   true,
}

// collectRefs walks a field value tree and appends every _ref alias found.
func collectRefs(value any, out *[]string) {
	switch v := value.(type) {
	case map[string]any:
		if ref, ok := v["_ref"].(string); ok {
			*out = append(*out, ref)
			return
		}
		for _, child := range v {
			collectRefs(child, out)
		}
	case []any:
		for _, child := range v {
			collectRefs(child, out)
		}
	}
}

// resolveRefs replaces each {"_ref": alias} with its temp id.
func resolveRefs(value any, aliasToTempID map[string]string) any {
	switch v := value.(type) {
	case map[string]any:
		if ref, ok := v["_ref"].(string); ok {
			if real, exists := aliasToTempID[ref]; exists {
				return real
			}
			return v
		}
		out := make(map[string]any, len(v))
		for k, child := range v {
			out[k] = resolveRefs(child, aliasToTempID)
		}
		return out
	case []any:
		out := make([]any, len(v))
		for i, child := range v {
			out[i] = resolveRefs(child, aliasToTempID)
		}
		return out
	default:
		return value
	}
}

// ResolvePayloadTree topo-sorts a create payload into an ordered list of CreateOp.
//
// create is the dashboard's nested map {model: [entity, ...]}.
// Each entity is a dict; _alias (declared by dependency targets) and _ref
// (declared by dependents, anywhere in the field tree) are the only reserved keys.
//
// Returns an error if the payload references an alias that is never declared,
// or if the alias graph contains a cycle.
func ResolvePayloadTree(create map[string]any) (*ResolvedTree, error) {
	// First pass: assign temp ids and collect alias declarations.
	type rawEntry struct {
		model  string
		tempID string
		entity map[string]any
		alias  string // "" if none
	}

	var entries []rawEntry
	counter := 0
	aliases := make(map[string]string)
	aliasOwnerModel := make(map[string]string)

	for model, entitiesRaw := range create {
		entities, ok := entitiesRaw.([]any)
		if !ok {
			return nil, ErrInvalidBody(fmt.Sprintf("`create.%s` must be a list of entity objects, got %T", model, entitiesRaw))
		}
		for _, entityRaw := range entities {
			entity, ok := entityRaw.(map[string]any)
			if !ok {
				return nil, ErrInvalidBody(fmt.Sprintf("`create.%s` entries must be objects, got %T", model, entityRaw))
			}
			tempID := fmt.Sprintf("__temp_%s_%d", model, counter)
			counter++

			alias := ""
			if a, hasAlias := entity["_alias"]; hasAlias {
				aStr, ok := a.(string)
				if !ok {
					return nil, ErrInvalidBody(`"_alias" must be a string`)
				}
				if _, dup := aliases[aStr]; dup {
					return nil, ErrInvalidBody(fmt.Sprintf(`duplicate _alias "%s"`, aStr))
				}
				aliases[aStr] = tempID
				aliasOwnerModel[aStr] = model
				alias = aStr
			}

			entries = append(entries, rawEntry{
				model:  model,
				tempID: tempID,
				entity: entity,
				alias:  alias,
			})
		}
	}

	// Second pass: collect each entry's dependency aliases and strip reserved keys.
	depsByTempID := make(map[string][]string)
	fieldsByTempID := make(map[string]map[string]any)
	modelByTempID := make(map[string]string)

	for _, entry := range entries {
		var deps []string
		cleaned := make(map[string]any)
		for key, value := range entry.entity {
			if reservedKeys[key] {
				continue
			}
			collectRefs(value, &deps)
			cleaned[key] = resolveRefs(value, aliases)
		}

		// Validate all refs point to known aliases.
		var unknown []string
		seen := make(map[string]bool)
		for _, a := range deps {
			if _, exists := aliases[a]; !exists && !seen[a] {
				unknown = append(unknown, a)
				seen[a] = true
			}
		}
		if len(unknown) > 0 {
			sort.Strings(unknown)
			return nil, ErrInvalidBody(fmt.Sprintf(
				"`create.%s` references unknown alias(es): %s",
				entry.model, strings.Join(unknown, ", ")))
		}

		depsByTempID[entry.tempID] = deps
		fieldsByTempID[entry.tempID] = cleaned
		modelByTempID[entry.tempID] = entry.model
	}

	// Build the temp_id graph and topo-sort.
	inDegree := make(map[string]int, len(entries))
	for _, e := range entries {
		inDegree[e.tempID] = 0
	}
	edges := make(map[string][]string) // dep -> []dependent

	for _, entry := range entries {
		seen := make(map[string]bool)
		for _, depAlias := range depsByTempID[entry.tempID] {
			depTempID := aliases[depAlias]
			if depTempID == entry.tempID || seen[depTempID] {
				continue
			}
			seen[depTempID] = true
			edges[depTempID] = append(edges[depTempID], entry.tempID)
			inDegree[entry.tempID]++
		}
	}

	// Kahn's, preserving payload order as the stable tie-breaker.
	payloadOrder := make(map[string]int, len(entries))
	for idx, e := range entries {
		payloadOrder[e.tempID] = idx
	}

	var ready []string
	for _, e := range entries {
		if inDegree[e.tempID] == 0 {
			ready = append(ready, e.tempID)
		}
	}
	sort.Slice(ready, func(i, j int) bool {
		return payloadOrder[ready[i]] < payloadOrder[ready[j]]
	})

	var sortedTempIDs []string
	for len(ready) > 0 {
		tid := ready[0]
		ready = ready[1:]
		sortedTempIDs = append(sortedTempIDs, tid)
		for _, nxt := range edges[tid] {
			inDegree[nxt]--
			if inDegree[nxt] == 0 {
				ready = append(ready, nxt)
			}
		}
		sort.Slice(ready, func(i, j int) bool {
			return payloadOrder[ready[i]] < payloadOrder[ready[j]]
		})
	}

	if len(sortedTempIDs) != len(entries) {
		var cycleModels []string
		for _, e := range entries {
			if inDegree[e.tempID] > 0 {
				cycleModels = append(cycleModels, modelByTempID[e.tempID])
			}
		}
		return nil, ErrInvalidBody(fmt.Sprintf(
			"cycle detected in _alias/_ref graph: %s",
			strings.Join(cycleModels, ", ")))
	}

	// Build CreateOp list in topo order.
	tree := &ResolvedTree{
		Aliases:           aliases,
		AliasOwnerModel:   aliasOwnerModel,
		AliasDependencies: make(map[string][]string),
	}
	for alias := range aliases {
		tree.AliasDependencies[alias] = depsByTempID[aliases[alias]]
	}
	for _, tid := range sortedTempIDs {
		tree.Ops = append(tree.Ops, CreateOp{
			Model:  modelByTempID[tid],
			Fields: fieldsByTempID[tid],
			TempID: tid,
		})
	}
	return tree, nil
}

// ComputeTeardownOrder orders models for teardown.
//
// With aliasDependencies available (newer refs tokens carry it), we run
// Kahn's topo sort over models and return the reverse topo so children
// are torn down before parents.
//
// Without it (older refs tokens), fall back to reversing the insertion
// order of refs keys (using modelOrder if available).
func ComputeTeardownOrder(
	refs map[string][]map[string]any,
	aliasDeps map[string][]string,
	aliasOwnerModel map[string]string,
	modelOrder []string,
) []string {
	// Use modelOrder if available (preserves creation-time ordering).
	// Otherwise fall back to map keys (non-deterministic in Go).
	var models []string
	if len(modelOrder) > 0 {
		// Use modelOrder but ensure all refs keys are included.
		seen := make(map[string]bool)
		for _, m := range modelOrder {
			if _, ok := refs[m]; ok {
				models = append(models, m)
				seen[m] = true
			}
		}
		for m := range refs {
			if !seen[m] {
				models = append(models, m)
			}
		}
	} else {
		for m := range refs {
			models = append(models, m)
		}
	}

	if len(aliasDeps) == 0 || len(aliasOwnerModel) == 0 {
		// Reverse insertion order.
		reversed := make([]string, len(models))
		for i, m := range models {
			reversed[len(models)-1-i] = m
		}
		return reversed
	}

	// Build model -> {model dependencies} by aggregating per-alias edges.
	modelSet := make(map[string]bool, len(models))
	for _, m := range models {
		modelSet[m] = true
	}
	modelDeps := make(map[string]map[string]bool, len(models))
	for _, m := range models {
		modelDeps[m] = make(map[string]bool)
	}

	for alias, deps := range aliasDeps {
		owner, ok := aliasOwnerModel[alias]
		if !ok || !modelSet[owner] {
			continue
		}
		for _, depAlias := range deps {
			depModel, ok := aliasOwnerModel[depAlias]
			if !ok || depModel == owner || !modelSet[depModel] {
				continue
			}
			modelDeps[owner][depModel] = true
		}
	}

	// Kahn's over models.
	inDegree := make(map[string]int, len(models))
	adj := make(map[string][]string, len(models))
	for _, m := range models {
		inDegree[m] = 0
	}
	for owner, deps := range modelDeps {
		for depModel := range deps {
			adj[depModel] = append(adj[depModel], owner)
			inDegree[owner]++
		}
	}

	payloadOrder := make(map[string]int, len(models))
	for i, m := range models {
		payloadOrder[m] = i
	}

	var ready []string
	for _, m := range models {
		if inDegree[m] == 0 {
			ready = append(ready, m)
		}
	}
	sort.Slice(ready, func(i, j int) bool {
		return payloadOrder[ready[i]] < payloadOrder[ready[j]]
	})

	var upOrder []string
	for len(ready) > 0 {
		m := ready[0]
		ready = ready[1:]
		upOrder = append(upOrder, m)
		for _, nxt := range adj[m] {
			inDegree[nxt]--
			if inDegree[nxt] == 0 {
				ready = append(ready, nxt)
			}
		}
		sort.Slice(ready, func(i, j int) bool {
			return payloadOrder[ready[i]] < payloadOrder[ready[j]]
		})
	}

	if len(upOrder) != len(models) {
		// Shouldn't happen (cycles rejected at up). Fall back.
		reversed := make([]string, len(models))
		for i, m := range models {
			reversed[len(models)-1-i] = m
		}
		return reversed
	}

	// Reverse the up order for teardown.
	reversed := make([]string, len(upOrder))
	for i, m := range upOrder {
		reversed[len(upOrder)-1-i] = m
	}
	return reversed
}

File: sdks/elixir/lib/autonoma/payload_topo.ex
defmodule Autonoma.PayloadTopo do
  @moduledoc """
  Resolve the create payload into an ordered list of operations.

  The create payload contains complete dependency information via `_alias`
  and `_ref` markers. Walking the payload to collect alias -> owner and
  owner -> {refs} gives the exact dependency graph. Kahn's topo sort over
  that graph produces the `up` order; the reverse is the `down` order.
  """

  alias Autonoma.Error

  @reserved_keys MapSet.new(["_alias", "_ref"])

  # ---------------------------------------------------------------------------
  # Walking the payload
  # ---------------------------------------------------------------------------

  @doc false
  def collect_refs(value, out) when is_map(value) do
    case Map.get(value, "_ref") do
      ref when is_binary(ref) ->
        [ref | out]

      _ ->
        Enum.reduce(value, out, fn {_k, v}, acc ->
          collect_refs(v, acc)
        end)
    end
  end

  def collect_refs(value, out) when is_list(value) do
    Enum.reduce(value, out, fn v, acc -> collect_refs(v, acc) end)
  end

  def collect_refs(_value, out), do: out

  @doc false
  def resolve_refs(value, alias_to_temp_id) when is_map(value) do
    case Map.get(value, "_ref") do
      ref when is_binary(ref) ->
        case Map.get(alias_to_temp_id, ref) do
          nil -> value
          real -> real
        end

      _ ->
        Map.new(value, fn {k, v} -> {k, resolve_refs(v, alias_to_temp_id)} end)
    end
  end

  def resolve_refs(value, alias_to_temp_id) when is_list(value) do
    Enum.map(value, fn v -> resolve_refs(v, alias_to_temp_id) end)
  end

  def resolve_refs(value, _alias_to_temp_id), do: value

  # ---------------------------------------------------------------------------
  # Tree resolution
  # ---------------------------------------------------------------------------

  @doc """
  Topo-sort a create payload into an ordered list of create ops.

  `create` is the dashboard's nested map `%{model => [entity, ...]}`.
  Each entity is a map; `_alias` (declared by dependency targets) and
  `_ref` (declared by dependents, anywhere in the field tree) are the
  only reserved keys.

  Returns `%{ops, aliases, alias_owner_model, alias_dependencies}`.
  """
  def resolve_payload_tree(create) when is_map(create) do
    # First pass: assign temp ids and collect alias declarations.
    {raw_entries, aliases, alias_owner_model} = first_pass(create)

    # Second pass: collect each entry's dependency aliases and strip reserved keys.
    {deps_by_temp_id, fields_by_temp_id, model_by_temp_id} =
      second_pass(raw_entries, aliases)

    # Build the temp_id graph and topo-sort.
    sorted_temp_ids = topo_sort(raw_entries, deps_by_temp_id, aliases, model_by_temp_id)

    # Build alias_dependencies
    alias_dependencies =
      Map.new(aliases, fn {alias_name, temp_id} ->
        deps = Map.get(deps_by_temp_id, temp_id, [])
        {alias_name, deps}
      end)

    # Build CreateOp list in topo order.
    ops =
      Enum.map(sorted_temp_ids, fn tid ->
        %{
          model: Map.fetch!(model_by_temp_id, tid),
          fields: Map.fetch!(fields_by_temp_id, tid),
          temp_id: tid
        }
      end)

    %{
      ops: ops,
      aliases: aliases,
      alias_owner_model: alias_owner_model,
      alias_dependencies: alias_dependencies
    }
  end

  def resolve_payload_tree(_create) do
    raise Error.invalid_body("`create` must be an object keyed by model name")
  end

  defp first_pass(create) do
    {raw_entries, _counter, aliases, alias_owner_model} =
      Enum.reduce(create, {[], 0, %{}, %{}}, fn {model, entities}, {entries, counter, aliases, aom} ->
        unless is_list(entities) do
          raise Error.invalid_body(
                  "`create.#{model}` must be a list of entity objects, got #{inspect(entities)}"
                )
        end

        Enum.reduce(entities, {entries, counter, aliases, aom}, fn entity, {entries, counter, aliases, aom} ->
          unless is_map(entity) do
            raise Error.invalid_body(
                    "`create.#{model}` entries must be objects, got #{inspect(entity)}"
                  )
          end

          temp_id = "__temp_#{model}_#{counter}"
          alias_val = Map.get(entity, "_alias")

          {aliases, aom, alias_str} =
            cond do
              is_binary(alias_val) ->
                if Map.has_key?(aliases, alias_val) do
                  raise Error.invalid_body("duplicate _alias \"#{alias_val}\"")
                end

                {Map.put(aliases, alias_val, temp_id), Map.put(aom, alias_val, model), alias_val}

              is_nil(alias_val) ->
                {aliases, aom, nil}

              true ->
                raise Error.invalid_body("\"_alias\" must be a string")
            end

          entry = {model, temp_id, entity, alias_str}
          {entries ++ [entry], counter + 1, aliases, aom}
        end)
      end)

    {raw_entries, aliases, alias_owner_model}
  end

  defp second_pass(raw_entries, aliases) do
    Enum.reduce(raw_entries, {%{}, %{}, %{}}, fn {model, temp_id, entity, _alias}, {deps_map, fields_map, model_map} ->
      {deps, cleaned} =
        Enum.reduce(entity, {[], %{}}, fn {key, value}, {deps, cleaned} ->
          if MapSet.member?(@reserved_keys, key) do
            {deps, cleaned}
          else
            new_deps = collect_refs(value, deps)
            resolved_value = resolve_refs(value, aliases)
            {new_deps, Map.put(cleaned, key, resolved_value)}
          end
        end)

      # Check for unknown aliases
      unknown = Enum.filter(deps, fn a -> not Map.has_key?(aliases, a) end)

      if unknown != [] do
        unique_unknown = unknown |> Enum.uniq() |> Enum.sort() |> Enum.join(", ")

        raise Error.invalid_body(
                "`create.#{model}` references unknown alias(es): #{unique_unknown}"
              )
      end

      {
        Map.put(deps_map, temp_id, deps),
        Map.put(fields_map, temp_id, cleaned),
        Map.put(model_map, temp_id, model)
      }
    end)
  end

  defp topo_sort(raw_entries, deps_by_temp_id, aliases, model_by_temp_id) do
    # Build payload order map for stable tie-breaking
    payload_order =
      raw_entries
      |> Enum.with_index()
      |> Map.new(fn {{_model, temp_id, _entity, _alias}, idx} -> {temp_id, idx} end)

    # Build in-degree and adjacency
    in_degree = Map.new(raw_entries, fn {_model, temp_id, _entity, _alias} -> {temp_id, 0} end)

    {in_degree, edges} =
      Enum.reduce(deps_by_temp_id, {in_degree, %{}}, fn {temp_id, deps}, {in_deg, adj} ->
        seen = MapSet.new()

        Enum.reduce(deps, {in_deg, adj, seen}, fn dep_alias, {in_deg, adj, seen} ->
          dep_temp_id = Map.fetch!(aliases, dep_alias)

          if dep_temp_id == temp_id or MapSet.member?(seen, dep_temp_id) do
            {in_deg, adj, seen}
          else
            seen = MapSet.put(seen, dep_temp_id)
            adj = Map.update(adj, dep_temp_id, [temp_id], fn existing -> existing ++ [temp_id] end)
            in_deg = Map.update!(in_deg, temp_id, fn d -> d + 1 end)
            {in_deg, adj, seen}
          end
        end)
        |> then(fn {in_deg, adj, _seen} -> {in_deg, adj} end)
      end)

    # Kahn's algorithm, preserving payload order as stable tie-breaker
    ready =
      in_degree
      |> Enum.filter(fn {_tid, deg} -> deg == 0 end)
      |> Enum.map(fn {tid, _} -> tid end)
      |> Enum.sort_by(fn t -> Map.get(payload_order, t, 0) end)

    sorted = kahns_loop(ready, edges, in_degree, payload_order, [])

    if length(sorted) != map_size(payload_order) do
      cycle_tids =
        in_degree
        |> Enum.filter(fn {_tid, deg} -> deg > 0 end)
        |> Enum.sort_by(fn {t, _} -> Map.get(payload_order, t, 0) end)
        |> Enum.map(fn {t, _} -> t end)

      cycle_models = Enum.map_join(cycle_tids, ", ", fn t -> Map.fetch!(model_by_temp_id, t) end)
      raise Error.invalid_body("cycle detected in _alias/_ref graph: #{cycle_models}")
    end

    sorted
  end

  defp kahns_loop([], _edges, _in_degree, _payload_order, sorted), do: sorted

  defp kahns_loop([tid | rest], edges, in_degree, payload_order, sorted) do
    neighbors = Map.get(edges, tid, [])

    {new_ready, in_degree} =
      Enum.reduce(neighbors, {[], in_degree}, fn nxt, {new_ready, in_deg} ->
        new_val = Map.get(in_deg, nxt, 1) - 1
        in_deg = Map.put(in_deg, nxt, new_val)

        if new_val == 0 do
          {[nxt | new_ready], in_deg}
        else
          {new_ready, in_deg}
        end
      end)

    ready =
      (rest ++ new_ready)
      |> Enum.sort_by(fn t -> Map.get(payload_order, t, 0) end)

    kahns_loop(ready, edges, in_degree, payload_order, sorted ++ [tid])
  end

  # ---------------------------------------------------------------------------
  # Teardown ordering
  # ---------------------------------------------------------------------------

  @doc """
  Order models for teardown.

  With `alias_dependencies` available (newer refs tokens carry it),
  we run the same Kahn's topo sort over models and return the reverse
  topo so children are torn down before parents.

  Without it (older refs tokens), fall back to reversing the insertion
  order of `refs` keys.
  """
  def compute_teardown_order(refs, alias_dependencies, alias_owner_model)
      when is_map(refs) do
    models = Map.keys(refs)

    if is_nil(alias_dependencies) or alias_dependencies == %{} or
         is_nil(alias_owner_model) or alias_owner_model == %{} do
      Enum.reverse(models)
    else
      # Build model -> {model dependencies}
      model_deps =
        Map.new(models, fn m -> {m, MapSet.new()} end)

      model_deps =
        Enum.reduce(alias_dependencies, model_deps, fn {alias_name, deps}, model_deps ->
          owner = Map.get(alias_owner_model, alias_name)

          if is_nil(owner) or not Map.has_key?(model_deps, owner) do
            model_deps
          else
            Enum.reduce(deps, model_deps, fn dep_alias, model_deps ->
              dep_model = Map.get(alias_owner_model, dep_alias)

              if is_nil(dep_model) or dep_model == owner or not Map.has_key?(model_deps, dep_model) do
                model_deps
              else
                Map.update!(model_deps, owner, fn s -> MapSet.put(s, dep_model) end)
              end
            end)
          end
        end)

      # Kahn's over models
      payload_order = models |> Enum.with_index() |> Map.new()

      in_degree = Map.new(models, fn m -> {m, 0} end)
      adj = %{}

      {in_degree, adj} =
        Enum.reduce(model_deps, {in_degree, adj}, fn {owner, deps}, {in_deg, adj} ->
          Enum.reduce(deps, {in_deg, adj}, fn dep_model, {in_deg, adj} ->
            adj = Map.update(adj, dep_model, [owner], fn existing -> existing ++ [owner] end)
            in_deg = Map.update!(in_deg, owner, fn d -> d + 1 end)
            {in_deg, adj}
          end)
        end)

      ready =
        in_degree
        |> Enum.filter(fn {_m, d} -> d == 0 end)
        |> Enum.map(fn {m, _} -> m end)
        |> Enum.sort_by(fn m -> Map.get(payload_order, m, 0) end)

      up_order = kahns_model_loop(ready, adj, in_degree, payload_order, [])

      if length(up_order) != length(models) do
        # Shouldn't happen - cycles rejected at up. Fall back.
        Enum.reverse(models)
      else
        Enum.reverse(up_order)
      end
    end
  end

  defp kahns_model_loop([], _adj, _in_degree, _payload_order, sorted), do: sorted

  defp kahns_model_loop([m | rest], adj, in_degree, payload_order, sorted) do
    neighbors = Map.get(adj, m, [])

    {new_ready, in_degree} =
      Enum.reduce(neighbors, {[], in_degree}, fn nxt, {new_ready, in_deg} ->
        new_val = Map.get(in_deg, nxt, 1) - 1
        in_deg = Map.put(in_deg, nxt, new_val)

        if new_val == 0 do
          {[nxt | new_ready], in_deg}
        else
          {new_ready, in_deg}
        end
      end)

    ready =
      (rest ++ new_ready)
      |> Enum.sort_by(fn m -> Map.get(payload_order, m, 0) end)

    kahns_model_loop(ready, adj, in_degree, payload_order, sorted ++ [m])
  end
end
